// http-proxy-lite/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const http = __importStar(require("http"));
const url_1 = require("url");
const events_1 = require("events");
class ProxyServer extends events_1.EventEmitter {
    web(req, res, options) {
        const emitter = new events_1.EventEmitter();
        const { hostname, port } = url_1.parse(options.target);
        const config = {
            hostname,
            port,
            path: req.url,
            method: req.method,
            headers: req.headers
        };
        if (req.method === 'POST') {
            const arr = [];
            req.on('data', chunk => {
                arr.push(chunk);
            });
            req.on('end', () => {
                let body = Buffer.concat(arr).toString();
                this.delegate(res, config, options, emitter, body);
            });
        }
        else {
            this.delegate(res, config, options, emitter);
        }
        return emitter;
    }
    delegate(rawRes, config, options, emitter, body) {
        let proxyReq = http.request(config, proxyRes => {
            rawRes.statusCode = proxyRes.statusCode || 500;
            for (let header in proxyRes.headers) {
                rawRes.setHeader(header, proxyRes.headers[header]);
            }
            proxyRes.on('data', chunk => {
                rawRes.write(chunk);
            });
            proxyRes.on('end', () => {
                rawRes.end();
                process.nextTick(() => {
                    const obj = { target: options.target };
                    this.listenerCount('end') && this.emit('end', obj);
                    emitter.listenerCount('end') && emitter.emit('end', obj);
                });
            });
        });
        proxyReq.on('error', e => {
            process.nextTick(() => {
                const obj = Object.assign({ target: options.target }, e);
                this.listenerCount('error') && this.emit('error', obj);
                emitter.listenerCount('error') && emitter.emit('error', obj);
            });
        });
        body && proxyReq.write(body);
        proxyReq.end();
    }
}
exports.createProxyServer = (opts) => new ProxyServer();
